packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -620,13 +549,30 @@ add (const Scalar& alpha,
      const Teuchos::RCP<const Map<LocalOrdinal, GlobalOrdinal, Node> >& rangeMap,
      const Teuchos::RCP<Teuchos::ParameterList>& params)
 {
-  typedef Map<LocalOrdinal,GlobalOrdinal,Node>                     map_type;
+  using Teuchos::RCP;
+  using Teuchos::rcpFromRef;
+  using Teuchos::rcp;
+  using Teuchos::ParameterList;
   typedef CrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>  crs_matrix_type;
-  Teuchos::RCP<const map_type> CrowMap = transposeB ?  B.getDomainMap() : B.getRowMap();
-
-  Teuchos::RCP<crs_matrix_type> C = rcp(new crs_matrix_type(CrowMap, 0));
-
-  add(alpha,transposeA,A,beta,transposeB,B,*C,domainMap,rangeMap,params);
+  //If transposeB, must compute B's explicit transpose to
+  //get the correct row map for C.
+  RCP<const crs_matrix_type> Brcp = rcpFromRef(B);
+  if(transposeB)
+  {
+    RowMatrixTransposer<Scalar, LocalOrdinal, GlobalOrdinal, Node> transposer(Brcp);
+    RCP<ParameterList> transposeParams(new ParameterList);
+    transposeParams->set ("sort", false);
+    Brcp = transposer.createTranspose(transposeParams);
+  }
+  //Check that A,B are fillComplete before getting B's column map
+  TEUCHOS_TEST_FOR_EXCEPTION
+    (! A.isFillComplete () || ! Brcp->isFillComplete (), std::invalid_argument,
+     "TpetraExt::MatrixMatrix::add(): A and B must both be fill complete.");
+  Teuchos::RCP<crs_matrix_type> C = rcp(new crs_matrix_type(Brcp->getRowMap(), 0));
+  //this version of add() always fill completes the result, no matter what is in params on input
+  if(!params.is_null())
+    params->set("Call fillComplete", true);
+  add(alpha, transposeA, A, beta, false, *Brcp, *C, domainMap, rangeMap, params);
```

Comment:
I'm guessing the point is that the changes to `add` below make this call work even if C is StaticProfile, as it would be with deprecated code in Tpetra disabled.